src/librustc_trait_selection/traits/error_reporting/suggestions.rs

```diff
@@ -748,20 +745,20 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                     return;
                 }
 
-                let trait_type = match mutability {
+                let suggested_ty = match mutability {
                     hir::Mutability::Mut => self.tcx.mk_imm_ref(region, t_type),
                     hir::Mutability::Not => self.tcx.mk_mut_ref(region, t_type),
                 };
 
-                let new_obligation = self.mk_obligation_for_def_id(
-                    trait_ref.skip_binder().def_id,
-                    trait_type,
-                    ObligationCause::dummy(),
+                let new_obligation = self.mk_trait_obligation_with_new_self_ty(
                     obligation.param_env,
+                    &trait_ref,
+                    suggested_ty,
```

Comment:
as above I don't 100% understand where this type is coming from or how we can be sure that it has no bound references in it -- doesn't it come from the `trait_ref`?

Reply:
`PolyTraitRef::self_ty` is equivalent to `self.skip_binder().substs.type_at(0)`. It seems we are assuming that the self type has no escaping bound variables. Is this assumption correct?

Reply:
One other thing that I found confusing: According to the [docs for `TraitRef`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TraitRef.html),

> [it] introduces a level of region binding, to account for higher-ranked trait bounds like `T: for<'a> Foo<&'a U>` or higher-ranked object types.

If that's true, why do we use `PolyTraitRef` in so many places? Is it just for higher-ranked self types? If so, why is `PolyTraitRef::self_ty`, which calls `skip_binder`, correct?

Reply:
These docs are wrong. Binders are always represented explicitly by a `Binder` -- so `TraitRef` alone cannot account for `for<'a>`, only a `PolyTraitRef` can.

Reply:
> Is this assumption correct?

I think this is sort of what I was asking. I doubt it's correct. I think a reasonable thing to do might be to *test* the assumption by using `no_bound_vars` instead of relying on `skip_binder` (and, if the assumption is false, not generating the suggestions).

Reply:
It seems like we want to change `PolyTraitRef::self_ty` to return a `Binder<Ty>`. Is this correct? This will have a significant amount of fallout I think. Do you want me to open an issue to track this? I could check the result of `self_ty` for `no_bound_vars` in this PR, but we need to be doing this across all diagnostics code, not just here.